doc/specs/#1595 - Suggestions UI/Snippets.md

```diff
@@ -0,0 +1,588 @@
+<kbd>enter</kbd> between them.
+
+```jsonc
+{
+    "command":
+    {
+        "action": "sendInput",
+        "input":
+        [
+            "$s=Invoke-GitHubGraphQlApi \"query{organization(login:`\"Microsoft`\"){projectV2(number: 159) { id } } }\"",
+            "$tasks = get-GitHubIssue  -Labels \"Issue-Task\" -state open",
+            "$bugs = get-GitHubIssue  -Labels \"Issue-Bug\" -state open",
+            "$issues = $tasks + $bugs",
+            "$issues | ? {$_.labels.Name -NotContains \"Needs-Triage\" } | ? { $_.milestone.title -Ne \"Icebox ❄\" } | ? type -Ne \"PullRequest\" | select -expand node_id | % {",
+            "  $resp = Add-GitHubBetaProjectItem -ProjectNodeId $s.organization.projectV2.id -ContentNodeId $_ ;",
+            "}",
+            ""
+        ]
+    },
+    "name": "Upload to project board",
+    "description": "Sync all our issues and bugs that have been triaged and are actually on the backlog to the big-ol project",
+},
+```
+
+This is slightly more maintainable. Assuming the user also has shell integration
+enabled, they could also set `"waitForSuccess": true`, and if any part of the
+script fails, then the rest of it won't be sent to the shell[[1](#footnote-1)].
+
+#### Fragment actions
+
+This was already added in [#16185]. These will allow third-party developers to
+create apps which add additional snippets to the Terminal. These will require
+app developers to add `id`s to each action they add in this way. Users can then
+bind that action `id` to a keybinding, if they so choose.
+
+Case in point:
+https://github.com/abduvik/just-enough-series/tree/master/courses/docker+docker-compose.
+Something like that should be able to be easily added directly to the Terminal.
+
+### Snippets pane
+
+With non-terminal content landing in 1.21 Preview, it's now simple to add
+additional types of panes to add to the Terminal. We'll support a new pane
+`"type": "snippets"`, to support opening the Snippets pane.
+
+This will be a pane with a `TreeView` in it and a text box to filter results
+(ala the Command Palette).
+
+Each item in the TreeView will be a kind of `FilteredCommand`, with a play
+button to support quickly running the command.
+
+This pane could also support all the different suggestion sources that the
+Suggestions UI supports - `recentCommands` could be plumbed into it from the
+currently active This pane could also support checkboxes to filter different
+suggestion sources.
+
+### Per-Project Snippets (`.wt.json`)
+
+> [INFO!]
+>
+> TODO!: Let's make sure to discuss the filename. It doesn't need to be
+> `.wt.json`. That seemed to match things like `.clang-format`, `.vsconfig`,
+> etc, but then also included the extension. However, node projects just use
+> `package.json` without the leading `.` for storing per-project commands.
+> Perhaps `.wt.json` is the worst of both? The best of both? `.wtconfig`?
+
+Users may also want to leave snippets in the root of their repo, for others to

```

Comment:
Does this require TermControl to have knowledge of TerminalSettingsModel, or jsoncpp? How does it generate (internally) SendInput actions? Is this a layering violation?

Reply:
"No"